core/include/scipp/core/variable.tcc

```diff
@@ -26,92 +26,55 @@ template <class T1, class T2> bool equal(const T1 &view1, const T2 &view2) {
-    const auto *xold = &*oldCoordT.values().begin();
-    const auto *xnew = &*newCoordT.values().begin();
-    // This function assumes that dimensions between coord and data either
-    // match, or coord is 1D.
-    const bool jointOld = oldCoordT.dims().shape().size() == 1;
-    const bool jointNew = newCoordT.dims().shape().size() == 1;
-#pragma omp parallel for
-    for (scipp::index c = 0; c < count; ++c) {
-      scipp::index iold = 0;
-      scipp::index inew = 0;
-      const scipp::index oldEdgeOffset = jointOld ? 0 : c * (oldSize + 1);
-      const scipp::index newEdgeOffset = jointNew ? 0 : c * (newSize + 1);
-      const auto oldOffset = c * oldSize;
-      const auto newOffset = c * newSize;
-      while ((iold < oldSize) && (inew < newSize)) {
-        auto xo_low = xold[oldEdgeOffset + iold];
-        auto xo_high = xold[oldEdgeOffset + iold + 1];
-        auto xn_low = xnew[newEdgeOffset + inew];
-        auto xn_high = xnew[newEdgeOffset + inew + 1];
-
-        if (xn_high <= xo_low)
-          inew++; /* old and new bins do not overlap */
-        else if (xo_high <= xn_low)
-          iold++; /* old and new bins do not overlap */
-        else {
-          // delta is the overlap of the bins on the x axis
-          auto delta = xo_high < xn_high ? xo_high : xn_high;
-          delta -= xo_low > xn_low ? xo_low : xn_low;
-
-          auto owidth = xo_high - xo_low;
-          newData[newOffset + inew] +=
-              oldData[oldOffset + iold] * delta / owidth;
-
-          if (xn_high > xo_high) {
-            iold++;
-          } else {
-            inew++;
-          }
-        }
-      }
-    }
-  }
-};
-
-template <typename T> struct RebinGeneralHelper {
-  static void rebin(const Dim dim, const Variable &oldT, Variable &newT,
-                    const Variable &oldCoordT, const Variable &newCoordT) {
-    const auto oldSize = oldT.dims()[dim];
-    const auto newSize = newT.dims()[dim];
-
-    const auto *xold = oldCoordT.values<T>().data();
-    const auto *xnew = newCoordT.values<T>().data();
-    // This function assumes that dimensions between coord and data
-    // coord is 1D.
-    int iold = 0;
-    int inew = 0;
+template <class T>
+// Special rebin version for rebinning inner dimension to a joint new coord.
+static void rebinInner(const Dim dim, const VariableConceptT<T> &oldT,
+                       VariableConceptT<T> &newT,
+                       const VariableConceptT<T> &oldCoordT,
+                       const VariableConceptT<T> &newCoordT, bool variances=false) {
+  scipp::span<const T> oldData;
+  scipp::span<T> newData;
+  if(variances) {
+    oldData = oldT.variances();
+    newData = newT.variances();
+  } else {
+    oldData = oldT.values();
+    newData = newT.values();
+  }
+  const auto oldSize = oldT.dims()[dim];
+  const auto newSize = newT.dims()[dim];
+  const auto count = oldT.dims().volume() / oldSize;
+  const auto *xold = &*oldCoordT.values().begin();
```

Comment:
Would `&oldCoordT.values()[0]` or even `const auto &xold = oldCoordT.values()` work here? Is `values()` supposed to return the `span`?

Reply:
Yes 👍, the pointer was some legacy from an attempt to handle cases with non-inner dimension. -> changed to use span directly.